File: nodejs/src/client/observe-facade.ts
/**
 * Observe facade (MCP: loxtep_observe).
 * Extends observe API with public queue read helpers and deployment status reads.
 */

import type { createObserveApi } from './observe.js';
import type { createQueuesApi } from './queues.js';
import type { DeploymentsApi } from './deployments.js';

export interface ObserveFacadeDeps {
  observe: ReturnType<typeof createObserveApi>;
  queues: ReturnType<typeof createQueuesApi>;
  deployments: DeploymentsApi;
}

export function createObserveFacade(deps: ObserveFacadeDeps): {
  status: ReturnType<typeof createObserveApi>['status'];
  stream_config: ReturnType<typeof createObserveApi>['stream_config'];
  get_queue_metadata: ReturnType<typeof createQueuesApi>['get_queue_metadata'];
  get_reader_checkpoint: ReturnType<typeof createQueuesApi>['get_reader_checkpoint'];
  open_reader: ReturnType<typeof createQueuesApi>['open_reader'];
  open_writer: ReturnType<typeof createQueuesApi>['open_writer'];
  list_deployments: DeploymentsApi['list'];
  get_deployment: DeploymentsApi['get'];
} {
  return {
    status: deps.observe.status.bind(deps.observe),
    stream_config: deps.observe.stream_config.bind(deps.observe),
    get_queue_metadata: deps.queues.get_queue_metadata.bind(deps.queues),
    get_reader_checkpoint: deps.queues.get_reader_checkpoint.bind(deps.queues),
    open_reader: deps.queues.open_reader.bind(deps.queues),
    open_writer: deps.queues.open_writer.bind(deps.queues),
    list_deployments: deps.deployments.list.bind(deps.deployments),
    get_deployment: deps.deployments.get.bind(deps.deployments),
  };
}

export type ObserveFacade = ReturnType<typeof createObserveFacade>;

File: python/src/loxtep/observe_facade.py
"""Observe facade (MCP: loxtep_observe)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Optional, Union

from .deployments import AsyncDeploymentsApi, DeploymentsApi


@dataclass
class ObserveFacade:
    _status: Callable[[], Any]
    _stream_config: Callable[[], Any]
    _get_queue_metadata: Callable[..., Any]
    _get_reader_checkpoint: Callable[..., Any]
    _open_reader: Callable[..., Any]
    _open_writer: Callable[..., Any]
    deployments: Optional[Union[DeploymentsApi, AsyncDeploymentsApi]] = None

    def status(self) -> Any:
        return self._status()

    def stream_config(self) -> Any:
        return self._stream_config()

    def get_queue_metadata(self, *args: Any, **kwargs: Any) -> Any:
        return self._get_queue_metadata(*args, **kwargs)

    def get_reader_checkpoint(self, *args: Any, **kwargs: Any) -> Any:
        return self._get_reader_checkpoint(*args, **kwargs)

    def open_reader(self, *args: Any, **kwargs: Any) -> Any:
        return self._open_reader(*args, **kwargs)

    def open_writer(self, *args: Any, **kwargs: Any) -> Any:
        return self._open_writer(*args, **kwargs)

    def list_deployments(self, **kwargs: Any) -> Any:
        if self.deployments is None:
            raise RuntimeError("deployments API is not configured on observe facade")
        return self.deployments.list(**kwargs)

    def get_deployment(self, deployment_id: str, **kwargs: Any) -> Any:
        if self.deployments is None:
            raise RuntimeError("deployments API is not configured on observe facade")
        return self.deployments.get(deployment_id, **kwargs)
